### PR TITLE
[SPARK-44910][SQL][3.4] Encoders.bean does not support superclasses with generic type arguments

### PR DESCRIPTION
**Pull request by gbloisi-openaire**
### What changes were proposed in this pull request?
This pull request adds Encoders.bean support for beans having a superclass declared with generic type arguments. 
For example:

```
class JavaBeanWithGenericsA<T> {
    public T getPropertyA() {
        return null;
    }

    public void setPropertyA(T a) {

    }
}

class JavaBeanWithGenericBase extends JavaBeanWithGenericsA<String> {
}

Encoders.bean(JavaBeanWithGenericBase.class); // Exception
```

That feature had to be part of [PR 42327](https://github.com/apache/spark/commit/1f5d78b5952fcc6c7d36d3338a5594070e3a62dd) but was missing as I was focusing on nested beans only (@hvanhovell )

### Why are the changes needed?
JavaTypeInference.encoderFor did not solve TypeVariable objects for superclasses so when managing a case like in the example above an exception was thrown.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Existing tests have been extended, new specific tests have been added

### Was this patch authored or co-authored using generative AI tooling?
No

@hvanhovell this is branch-3.4 port of [PR-42634](https://github.com/apache/spark/pull/42634)

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
cc @hvanhovell

File: sql/catalyst/src/test/java/org/apache/spark/sql/catalyst/JavaTypeInferenceBeans.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst;
+
+public class JavaTypeInferenceBeans {
+
+    static class JavaBeanWithGenericsA<T> {
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/test/java/org/apache/spark/sql/catalyst/JavaTypeInferenceBeans.java:22]: Although I understand that this following the style of the deleted `JavaBeanWithGenerics.java`. Shall we use 2-space indentation, if you don't mind?

Comment (gbloisi-openaire) [reply sql/catalyst/src/test/java/org/apache/spark/sql/catalyst/JavaTypeInferenceBeans.java]: No problem, I just committed the required change. I also updated [PR-42634](https://github.com/apache/spark/pull/42634), that is the same change for master and 3.5 branches

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM.

**Comment by dongjoon-hyun** [comment]
Merged to branch-3.4.

**Pull request closed without merging**